modules/aaudio/player.c

```diff
@@ -0,0 +1,173 @@
+/**
+ * @file player.c  Aaudio sound driver - player
+ *
+ */
+
+
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <string.h>
+
+#include <aaudio/AAudio.h>
+
+#include "aaudio.h"
+
+struct auplay_st {
+	struct auplay_prm play_prm;
+	auplay_write_h *wh;
+	size_t sampsz;
+	void *arg;
+};
+
+AAudioStream *playerStream = NULL;
+
+
+static void auplay_destructor(void *arg)
+{
+	struct auplay_st *st = arg;
+
+	if (playerStream)
+		AAudioStream_close(playerStream);
+
+	st->wh = NULL;
+}
+
+/**
+ * For an output stream, this function should render and write from
+ * userData numFrames of data in the streams current data format to
+ * the audioData buffer.
+ */
+ static int dataCallback(AAudioStream *stream, void *userData,
+			 void *audioData, int32_t numFrames) {
+	(void)stream;
+	struct auplay_st *st = userData;
+	struct auframe af;
+
+	auframe_init(&af, st->play_prm.fmt, audioData, numFrames,
+		     st->play_prm.srate, st->play_prm.ch);
+
+	st->wh(&af, st->arg);
+
+	for (size_t i = 0; i < af.sampc; ++i)
+		((int16_t *)audioData)[i] = ((int16_t *)(af.sampv))[i];
```

Comment:
But wait, this should not needed, since `af.sampv` and `audioData` is the same pointer.

```c
st->wh(&af, st->arg);
```

Already writes to the buffer. So you can simply drop the `for` loop

Reply:
Is it so that also stereo (2 channels) should be supported and other formats than S16LE should be supported?  I have never needed those in my Android app.

Regarding `memcpy`, the change didn't compile until I changed it to this:
```
	memcpy(audioData, af.sampv, auframe_size(&af) * af.ch);
```

Reply:
As said, you can remove the memcpy. Since `st->wh` writes directly to `audioData` here, since `af.sampc` and `audioData` are the same pointer, so it's currently copied twice.